// SINTAXIS PROTO3 PARA PODER TENER OBJETOS :)
syntax = "proto3";

package chat;

// REGISTRO DE USUARIOS NUEVOS
message UserRegistration {
    optional string username = 1;
    optional string ip = 2;
}


// INFORMACION DEL USUARIO BASE
message UserInfo {
    optional string username = 1;
    optional string status = 2;
    optional string ip = 3;
}


// SOLICITAR INFORMACION DE UN USUARIO O DE TODOS
// user: (username || everyone)
message UserRequest {
    optional string user = 1;
}


// ENVIADO POR EL SERVIDOR
message ConnectedUsersResponse {
    repeated UserInfo connectedUsers = 1;
}


// CAMBIAR DE ESTADO
// activo, inactivo, ocupado
message ChangeStatus {
    optional string username = 1;
    optional string status = 2;
}


// ENVIO DE MENSAJES (DIRECTOS || PA TODOS)
// recipient: (username || everyone)
// sender: username del remitente
message MessageCommunication {
    optional string message = 1;
    optional string recipient = 2; 
    optional string sender = 3; 
}


// option 1: Registro de Usuarios
// option 2: Usuarios Conectados
// option 3: Cambio de Estado
// option 4: Mensajes
// option 5: Informacion de un usuario en particular
(SON LAS MISMAS OPCIONES QUE LO QUE RESPONDERA EL SERVER PARA ESTAR EN EL MISMO FORMATO)
// PARA LA INFO DE TODOS LOS USUARIOS CONECTADOS MANDAR UN UserRequest con everyone
// PARA LA INFO DE SOLO 1 USUARIO MANDAR UN UserRequest con el username
message ClientPetition {
    optional int32 option = 1;

    optional UserRegistration registration = 2;

    optional UserRequest users = 3;

    optional ChangeStatus change = 4;

    optional MessageCommunication messageCommunication = 5;
}


// option 1: Registro de Usuarios
// option 2: Usuarios Conectados
// option 3: Cambio de Estado
// option 4: Mensajes
// option 5: Informacion de un usuario en particular
// codigo 200: Exito
// codigo 500: Fallo
// EN RESPUESTA A UNA SOLICITUD DEL CLIENTE POR EJEMPLO: REGISTRO DE USUARIOS
// EL SERVER RETORNA:
OPCION, 
CODIGO 
MENSAJE DEL SERVIDOR
// PARA SABER SI EL REQUEST FUE SUCCESS O ALGO MURIO EN EL SERVER 
message ServerResponse {
    optional int32 option = 1;

    optional int32 code = 2;

    optional string serverMessage = 3;

    optional ConnectedUsersResponse connectedUsers = 4;

    optional MessageCommunication messageCommunication = 5;

    optional UserInfo userInfoResponse = 6;

    optional ChangeStatus change = 7;
}


